hardware_integration/src/mcp2515_can_interface.cpp

```diff
@@ -0,0 +1,383 @@
+//================================================================================================
+/// @file mcp2515_can_interface.cpp
+///
+/// @brief An interface for using the MCP2515 can controller.
+/// @author Daan Steenbergen
+///
+/// @copyright 2022 Adrian Del Grosso
+//================================================================================================
+#include "isobus/hardware_integration/mcp2515_can_interface.hpp"
+#include "isobus/isobus/can_warning_logger.hpp"
+#include "isobus/utility/system_timing.hpp"
+
+#include <iostream>
+
+MCP2515CANInterface::MCP2515CANInterface(SPIHardwarePlugin *transactionHandler, const std::uint8_t cfg1, const std::uint8_t cfg2, const std::uint8_t cfg3) :
+  transactionHandler(transactionHandler),
+  cfg1(cfg1),
+  cfg2(cfg2),
+  cfg3(cfg3)
+{
+}
+
+MCP2515CANInterface::~MCP2515CANInterface()
+{
+	close();
+}
+
+bool MCP2515CANInterface::get_is_valid() const
+{
+	return (transactionHandler) && (initialized);
+}
+
+void MCP2515CANInterface::close()
+{
+	initialized = false;
+}
+
+void MCP2515CANInterface::open()
+{
+	if (reset())
+	{
+		if (set_mode(MCPMode::CONFIG))
+		{
+			if ((write_register(MCPRegister::CNF1, cfg1)) &&
+			    (write_register(MCPRegister::CNF2, cfg2)) &&
+			    (write_register(MCPRegister::CNF3, cfg3)))
+			{
+				if (set_mode(MCPMode::NORMAL))
+				{
+					initialized = true;
+				}
+			}
+		}
+	}
+}
+
+bool MCP2515CANInterface::reset()
+{
+	bool retVal = false;
+	if (write_reset())
+	{
+		// Depends on oscillator & capacitors used
+		std::this_thread::sleep_for(std::chrono::milliseconds(10));
+
+		std::uint8_t zeros[14] = { 0 };
+		if ((write_register(MCPRegister::TXB0CTRL, zeros, 14)) &&
+		    (write_register(MCPRegister::TXB1CTRL, zeros, 14)) &&
+		    (write_register(MCPRegister::TXB2CTRL, zeros, 14)) &&
+		    (write_register(MCPRegister::RXB0CTRL, 0)) &&
+		    (write_register(MCPRegister::RXB1CTRL, 0)) &&
+		    (write_register(MCPRegister::CANINTE, 0xA3)))
+		{
+			retVal = true;
+		}
+	}
+	return retVal;
+}
+
+bool MCP2515CANInterface::get_read_status(std::uint8_t &status)
+{
+	bool retVal = false;
+	if (transactionHandler)
+	{
+		SPIHardwarePlugin::TransactionFrame frame{
+			.txBuffer = { static_cast<std::uint8_t>(MCPInstruction::READ_STATUS), 0x00 },
```

Comment:
GCC complains about designated initializers here and elsewhere `C++ designated initializers only available with ‘-std=c++2a’ or ‘-std=gnu++2a’`

Perhaps it might make more sense to add a constructor to the struct or make it a class? Either one of those is maybe more portable.

Reply:
Agreed, added a `SPITransactionFrame` class in https://github.com/ad3154/ISO11783-CAN-Stack/pull/91/commits/5cdd9daa43de5360fe677c76b0cd392445602928